plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE GADTs            #-}
+{-# LANGUAGE TypeFamilies     #-}
+
+{-|
+Call site inlining machinery. For now there's only one part: inlining of fully applied functions.
+See ADR TODO for motivation. We inline fully applied functions if the cost and size are acceptable.
+See note [Inlining of fully applied functions].
+-}
+
+module PlutusIR.Transform.Inline.CallSiteInline where
+
+{- Note [Inlining of fully applied functions]
+
+We inline the call site when a function is fully applied AND the cost and size are acceptable.
+
+Consider `let v :: ty = rhs in body`, in which `body` calls `v`.
+
+We consider cases when `v` is a function/lambda abstraction(s). I.e.:
+
+let v :: type = \x1.\x2...\xn.VBody in body
+
+In the `body`, where `v` is *called*,
+if it was given `n` arguments, then it is fully applied in the `body`.
+We inline the *call site* of `v` in this case, i.e.,
+we replace `v x_1 ... x_n` in the `body` with `rhs`. E.g.
+
+let f :: TyLam (TODO)= \x.\y -> x*y
+    y :: TyLam = f q
+in f a b
+
+becomes `let y = f q in a*b`. What about
+
+let f = \x.\y -> x*y
+    y = f q
+in f a b + y
+
+Here we have a `f` that is not fully applied via `y`, so we cannot eliminate the let-binding of `f`.
+TODO
+To do this we need to count the number of type/term lambda abstractions,
+so we know how many term/type arguments they have.
+We also need to decide whether or not to inline based on the call site, which we don’t currently.
+
+The heuristic: Counting lambdas and arguments
+
+We check for a binding that is a lambda abstraction and is fully applied in the body.
+
+E.g.,
+
+
+We pattern match the _rhs_ with `LamAbs` or `TyAbs` (lambda abstraction for terms or types),
+and count the number of lambdas.
+Then, in the _body_, we check for term or type application (`Apply` or `TyInst`) of _v_.
+
+If _v_ is fully applied in the body, i.e., if
+
+1. the number of type lambdas equals the number of type arguments applied, and
+2. the number of term lambdas equals the number of term arguments applied, and
+
+if other call site inlining conditions are satisfied,
+  (we currently inline too little, this will be improved later, see below)
+
+we inline _v_, i.e., replace its occurrence with _rhs_ in the _body_.
+For the rest of the discussion here we focus on the conditions 1 and 2 only.
+
+E.g., for
+
+```haskell
+let f = \x.\y.x+y in
+    f 2 4
+```
+
+the _body_ `f 2 4` is a function application of `f`. Since `f` is fully applied, we inline `f`.
+
+### Examples
+
+Below are some examples that involve a _body_ that is not fully reducible/applied but
+following our heuristic is beneficial.
+
+#### Example 1: function in body
+
+```haskell
+let f = \x . x
+     g = \y . f
+in g a
+```
+
+`f` and `g` each has 1 lambda. However, `g`'s _body_ includes `f` which also has a lambda.
+Since we only count the number of lambdas, `g` is fully applied, and we inline.
+`g a` reduces to `f`, which reduces the amount of code even though `f` has a lambda that
+is not applied.
```

Comment:
The "even though `f` has a lambda that is not applied" part doesn't make sense to me. `f` is not fully applied, and we don't inline `f`, and we shouldn't inline `f`. I don't see a problem here.